[DPT-14866] Add explicit cache expiration.

anthonyjpratti opened the pull request:
This change was to make the groupcache ttl aware.
This was done through 3 changes.
1. Propagate ttl from the underlying data store, and the Get code paths in groupcache.
2. Store ttl for each value in both the hot and main cache.
3. If the ttl occurs before the current time, do not cache or retrieve cache.
    This has the effect of the expired resources being evicted from the lru cache.

anthonyjpratti committed:
Add explicit cache expiration.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/groupcache?pullRequest=8) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twitter/groupcache?pullRequest=8) before we can accept your contribution.<br/><hr/>**Anthony Pratti** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/groupcache?pullRequest=8) it.</sub>

ryancouto commented on groupcache.go:
```diff
@@ -247,36 +247,41 @@ func (g *Group) initPeers() {
+	return expiration, setSinkView(dest, value)
+}
+
+type loadResult struct {
```
could you add a comment here describing loadResult?

anthonyjpratti replied:
will do

anthonyjpratti replied:
loadResult originally was to match the interface of flightgroup.Do, but now it is generalized as a payload type for all functions that take or return value and expiration

ryancouto replied:
sorry, i wasn't totally clear - i meant a comment in the code

anthonyjpratti replied:
Oh you were clear, I was just adding context because loadResult is going away.

anthonyjpratti replied:
Done.

ryancouto commented on groupcache.go:
```diff
@@ -522,6 +537,30 @@ type cache struct {
+	metadata   map[string]*cacheValueMetadata
+}
+
+type cacheValueMetadata struct {
```
ditto re: comment here about cacheValueMetadata, why it's structured this way, what it is, etc.

anthonyjpratti replied:
will do

anthonyjpratti replied:
Done.

ryancouto commented on groupcache.go:
```diff
@@ -552,19 +593,22 @@ func (c *cache) add(key string, value ByteView) {
 		return
 	}
+	if expiration = c.getMetadata(key).expiration; expiration != nil && expiration.Before(time.Now().UTC()) {
+		return
```
should we remove it from the cache here?

anthonyjpratti replied:
The implementation of the LRU doesn't allow for explicit cache eviction, but it will eventually be evicted.

ryancouto commented on groupcache_test.go:
```diff
@@ -61,16 +62,18 @@ const (
 	cacheSize       = 1 << 20
 )
 
+var expiration = time.Now().UTC().Add(time.Hour)
```
can throw this into the var block declared above on line 40

anthonyjpratti replied:
will do

anthonyjpratti replied:
Done.

dgdonut commented on groupcache.go:
```diff
@@ -48,13 +48,13 @@ type Getter interface {
 	// uniquely describe the loaded data, without an implicit
```
Update comment to reflect expiration return

anthonyjpratti replied:
Done.

dgdonut commented on groupcache.go:
```diff
@@ -247,36 +247,41 @@ func (g *Group) initPeers() {
 
 // underlying Get logic - loads key either by invoking the getter locally or by sending it to another machine.
-func (g *Group) load(ctx Context, key string, dest Sink) (value ByteView, destPopulated bool, err error) {
+func (g *Group) load(ctx Context, key string, dest Sink) (value ByteView, destPopulated bool, expiration *time.Time, err error) {
```
style nit: return expiration 2nd, after value, to match other fns

dgdonut replied:
actually, why not just return a loadResult explicitly for all functions where we include the ByteView and *time.Time in the result?

anthonyjpratti replied:
I was inserting second to last, but I agree.
I'll change it to use loadResult everywhere (with a better name tho, payload maybe?)

anthonyjpratti replied:
Done.

dgdonut commented on groupcache.go:
```diff
@@ -416,7 +429,8 @@ func (g *Group) store(ctx Context, key string, data []byte, ttl time.Duration) (
 		g.Stats.LocalStores.Add(1)
 		value := ByteView{b: data}
-		g.populateCache(key, value, &g.mainCache)
+		expiration := time.Now().UTC().Add(ttl)
```
Hm. This makes the expiration time dependent on the store operation, which isn't ideal. I see why you had to do this - I added the ttl pass-through value previously as a Duration, which in retrospect was Wrong™, because getting a Time value from a Duration is dependent on the observer. In practice this value will likely end up being negligibly off from what the caller would interpret, but I think this is an important thing to correct now. Sorry that you to stumbled into this.

Does this make sense / any reason not to go this route? I think what we'd end up doing is changing the previous pass through values to time.Times and then not having to compute these values from Durations (and making the corresponding change in scoot afterwards as well).

anthonyjpratti replied:
Totally Agree. Will do.

anthonyjpratti replied:
Done.

dgdonut commented on groupcache.go:
```diff
@@ -443,30 +457,31 @@ func (g *Group) putFromPeer(ctx Context, peer ProtoPeer, key string, data []byte
 	// conditionally populate hotCache.  For now just do it some
 	// percentage of the time.
 	if rand.Intn(10) == 0 {
```
While we're in here change: make hot cache percentage a const somewhere

anthonyjpratti replied:
will do

anthonyjpratti replied:
Done.

dgdonut commented on groupcache.go:
```diff
@@ -552,19 +593,22 @@ func (c *cache) add(key string, value ByteView) {
 	if c.lru == nil {
 		return
 	}
+	if expiration = c.getMetadata(key).expiration; expiration != nil && expiration.Before(time.Now().UTC()) {
```
seems like getMetadata protects us against the nil check so it's not needed. nil deref defense isn't a terrible thing to be redundant about though, your preference on whether to leave it

anthonyjpratti replied:
getMetadata does protect us against the nil for the metadata itself, but expiration could still be nil.

dgdonut replied:
oh - yep

dgdonut commented on groupcache_test.go:
```diff
@@ -534,7 +539,7 @@ func TestNoDedup(t *testing.T) {
 		go func() {
 			var s string
-			if err := g.Get(dummyCtx, testkey, StringSink(&s)); err != nil {
+			if _, err := g.Get(dummyCtx, testkey, StringSink(&s)); err != nil {
```
Do we have an explicit test against the don't-serve-from-cache-if-expired case?

anthonyjpratti replied:
Nope, will add

anthonyjpratti replied:
I added, it did bring up an interesting case. If a getter itself returns a value with an expired ttl, the current implementation will return the value, and will effectively not cache it. Thoughts on this behavior?

dgdonut replied:
I think in that case we want to intercept that in getLocally and getFromPeer, and if ttl is not nil and has expired, return the same behavior as if it wasn't found.

anthonyjpratti replied:
Done.

anthonyjpratti committed:
[DPT-14866]
Add comments to cacheValueMetadata.
Encapsulate value and expiration in payload struct.

anthonyjpratti committed:
Rename expiration to ttl.
Change ttl from Duration to Time type.
Add test for resource expiration.
Extract populateHotCacheOdds into a const.

anthonyjpratti committed:
Ignore resources in Get code path if ttl is expired.

ryancouto commented on groupcache.go:
```diff
@@ -416,57 +436,62 @@ func (g *Group) store(ctx Context, key string, data []byte, ttl time.Duration) (
 	if err != nil {
 		return err
 	}
 	value := ByteView{b: data}
```
big nit: this could be pulled into the `if rand.Intn()...` block too

anthonyjpratti replied:
I try to avoid refactors like that in PR for Features, but I totally agree. I was debating it myself. Done.

anthonyjpratti replied:
You are referring to value := ByteView{b: data}, correct?

ryancouto replied:
yup! agreed in general, but it's a small enough change that I think it's ok

anthonyjpratti replied:
Totally agree, I honestly thought about doing it for like 5 minutes. Glad you mentioned this.

anthonyjpratti committed:
Move value into if statement for probabilistic cache population.
